### PR TITLE
[StaticLogic] Add new PipelineWhile operation.

### PR DESCRIPTION
**Pull request by mikeurbach**
https://llvm.discourse.group/t/rfc-representing-pipelined-loops/4171/1
for more discussion.

This defines the operation's syntax more or less as discussed,
including round-trip tests for the four running examples we've had.

In order to reduce test churn, the existing PipelineOp and the
StandardToStaticLogic pass are left untouched for now. Ultimately, the
new pipeline operation could replace the existing PipelineOp.

### CODE REVIEW

**Comment by mortbopet** [comment]
Looks good to me and summarizes nicely what we've been discussing over on discourse.
I am still wondering about passing values from different stages to the terminator `iter_args` and how this affects the hardware we expect to generate/II of the pipeline.
Say that we want to implement a pipeline that accumulates an array:
```
func @acc(%mem: memref<?xi32>) {
  %c0 = constant 0 : index
  %c1 = constant 1 : index
  %c10 = constant 10 : index
  staticlogic.pipeline.while iter_args(%iv = %c0, %acc = %c0) : (index, i32) -> () {
    %3 = cmpi ult, %iv, %c10 : index
    staticlogic.pipeline.yield %3 : i1
  } do {
    %3:2 = staticlogic.pipeline.stage  {
      %4 = addi %iv, %c1 : index
      %5 = memref.load %mem[%iv] : memref<1xi32>
      staticlogic.pipeline.yield %5, %6 : index, i32
    } : index, i32
    %4 = staticlogic.pipeline.stage  {
      %5 = addi %acc, %3#1 : i32
      staticlogic.pipeline.yield %5 : i32
    } : i32
    staticlogic.pipeline.terminator iter_args(%3#0, %4), results(%4) : (index, i32) -> (i32)
  }
  return
}
```
I'd hope to generate something like
image
What is notable here is:
- the initial value of `%acc` in the `iter_args` of the pipeline initializes a register value in the 2nd stage of the pipeline
- the feedback value `%4` is fed back to the 2nd stage of the pipeline.

Is this inferable from our representation? I think it is, and it most likely comes down to the difference between the stage difference between the use and yields of `iter_args`. I.e. in the above `%acc` is both used and defined in the 2nd stage.

For an example of use/def stage difference, let's modify the example above to use the accumulator value in the incrementation of the induction variable:
```
func @acc(%mem: memref<?xi32>) {
  %c0 = constant 0 : index
  %c1 = constant 1 : index
  %c10 = constant 10 : index
  staticlogic.pipeline.while iter_args(%iv = %c0, %acc = %c0) : (index, i32) -> () {
    %3 = cmpi ult, %iv, %c10 : index
    staticlogic.pipeline.yield %3 : i1
  } do {
    %3:2 = staticlogic.pipeline.stage  {
      %4 = addi %iv, %acc : index
      %5 = memref.load %mem[%iv] : memref<1xi32>
      staticlogic.pipeline.yield %5, %6 : index, i32
    } : index, i32
    %4 = staticlogic.pipeline.stage  {
      %5 = addi %acc, %3#1 : i32
      staticlogic.pipeline.yield %5 : i32
    } : i32
    staticlogic.pipeline.terminator iter_args(%3#0, %4), results(%4) : (index, i32) -> (i32)
  }
  return
}
```
which i imagine would look like this:
image

In this case, the .

This situation gets a lot more complicated in situations where there's **many** `iter_args` and their use/def locations are not "nicely" placed either at the start or end of the pipeline. I have to read up on the specifics, but i'd assume that the II of our loop is constrained by the maximum difference between any use/def in the pipeline, and this is then what will guide the register insertions _or_ `write_en` signals of the different stage registers in the pipeline.

**to sum up**; i'm still a supporter of this representation. There is a lot of implicit knowledge about the pipeline which needs to be inferred in a lowering pass, but i'm not sure that we will gain anything by trying to force all of this stuff into the IR, since we should be able to generate said knowledge from this representation.

**Comment by mikeurbach** [comment]
> the II of our loop is constrained by the maximum difference between any use/def in the pipeline

Exactly, this is my mental model as well. I can add some documentation about how the `iter_args` relate to II and the hardware this abstracts.

> There is a lot of implicit knowledge about the pipeline which needs to be inferred in a lowering pass, but i'm not sure that we will gain anything by trying to force all of this stuff into the IR, since we should be able to generate said knowledge from this representation.

Yep, I think this is the point of having such an IR. The semantics of this operation implicitly capture these things. My hope (and I think this is the goal of all IR design) is to define clear semantics that lead to a well-defined lowering, which makes these implicit rules explicit. Does that make sense? Are we capturing the right information at this level?

**Comment by jopperm** [comment]
> > the II of our loop is constrained by the maximum difference between any use/def in the pipeline
>
> Exactly, this is my mental model as well. I can add some documentation about how the `iter_args` relate to II and the hardware this abstracts.

I think a more explicit encoding of the II in form of an integer attribute would be useful in this representation, as the use/def distance is only one aspect constraining the II (the presence of shared operator instances/resource constraints is the other one). Or maybe a frontend wants to achieve a specific II to match throughput with other pipelined loops in a larger design.

File: include/circt/Dialect/StaticLogic/StaticLogic.td [hunk 1]
```diff
@@ -72,4 +72,137 @@ def ReturnOp : Op<StaticLogic_Dialect, "return", [Terminator]> {
...
+    for representing loop-carried values like induction variables or reductions.
+    When the pipeline starts execution, the registers indicated as `iter_args`
+    by `staticlogic.pipeline.terminator` should be initialized to the initial
+    values specified in the `iter_args` section here.
+
+    The single-block `condition` region dictates the condition under which the
+    pipeline should execute. It has a `staticlogic.yield` terminator, and the
+    pipeline executes while the yielded value is `true : i1`. It may access SSA
+    values dominating the pipeline, as well as `iter_args`, which are block
+    arguments.
```
Comment (jopperm) [inline include/circt/Dialect/StaticLogic/StaticLogic.td:96]: Do we expect/enforce the condition region to be combinatorial? I think that would be reasonable limitation, but then should be mentioned here.

**Comment by jopperm** [comment]
@mikeurbach What's your current mental model for multi-cycle operations? 

Assuming the `memref.load` in @mortbopet's first example takes 2 stages instead of one:
```mlir
...
  } do {
    %3:2 = staticlogic.pipeline.stage  {
      %4 = addi %iv, %c1 : index
      %5 = memref.load %mem[%iv] : memref<1xi32>
      staticlogic.pipeline.yield %4, %5 : index, i32
    } : index, i32
    %7 = staticlogic.pipeline.stage  {
      staticlogic.pipeline.yield %3#1 : i32
    } : i32
    %4 = staticlogic.pipeline.stage  {
      %5 = addi %acc, %7 : i32
      staticlogic.pipeline.yield %5 : i32
    } : i32
    staticlogic.pipeline.terminator iter_args(%3#0, %4), results(%4) : (index, i32) -> (i32)
  }
...
```
In principle, I'd say the proposed semantics work well here, because they allow to express an "empty" stage that just forwards a value. However, inside stage `%7`, the value `%3#1` represents merely a future value (the load unit hasn't returned its result yet), but nothing would stop you from inserting another user of it in stage `%7`. Not sure if this is a real problem, though.

File: include/circt/Dialect/StaticLogic/StaticLogic.td (hunk 1 above)
Comment (mikeurbach) [reply include/circt/Dialect/StaticLogic/StaticLogic.td:96]: Yes, that was my thinking. I'll document this, and the verifier can enforce it.

**Comment by mikeurbach** [comment]
> What's your current mental model for multi-cycle operations?

Hmm, that's a good question. Your proposal with an "empty" stage sort of makes sense, but I don't think yielding `%3#1` matches my mental model. In my head, each `staticlogic.pipeline.yield` actually marks values to be "registered" in pipeline registers. I was having a discussion about this with Morten, and it might make more sense to instead call it `staticlogic.pipeline.register`. The `yield` terminology was taken from SCF (along with `iter_args`, which could also be renamed if necessary).

Thinking about how I'd represent a multi-cycle operation, I do think an empty stage may be necessary, but in this case I wouldn't even specify the value in the terminator. Something like this: 

```mlir
  } do {
    %3:2 = staticlogic.pipeline.stage  {
      %4 = addi %iv, %c1 : index
      %5 = memref.load %mem[%iv] : memref<1xi32>
      staticlogic.pipeline.yield %4, %5 : index, i32
    } : index, i32
    %7 = staticlogic.pipeline.stage  {
      staticlogic.pipeline.yield
    }
    %4 = staticlogic.pipeline.stage  {
      %5 = addi %acc, %3#1 : i32
      staticlogic.pipeline.yield %5 : i32
    } : i32
    staticlogic.pipeline.terminator iter_args(%3#0, %4), results(%4) : (index, i32) -> (i32)
  }
...
```

This seems more in line with how I'm imagining the terminator to explicitly designate values to be registered, as well as my intent for each stage to represent the duration of one cycle. However, this breaks the invariant I initially proposed that each stage either reads `iter_args` or values registered by the previous stage. Perhaps this can just be relaxed: each stage should either read `iter_args` or values registered by _any_ previous stage. I think that would provide some guardrails to make how things feed forward and back explicit, while still admitting multi-cycle operations. Thoughts?

> I think a more explicit encoding of the II in form of an integer attribute would be useful in this representation

I think that makes sense. Morten and I discussed this, and he is actually working on representing HLS directives coming from the frontend as attributes, so we can probably use that here.

Somewhat related, we continued our discussion about II and iter_args. One interesting case Morten brought up is what happens when iter_args feedback into the middle of a pipeline. In rough ascii-art:

```
             ↓-------------\
|stage1| -> |stage2| -> |stage3|
```

(@mortbopet please correct me if I'm misrepresenting, this is what I drew in my notes)

I'm trying to work up an example of what that would look like (as well as the example I previously mentioned, for II>1). I'll come back with those examples for discussion.

**Comment by jopperm** [comment]
> In my head, each `staticlogic.pipeline.yield` actually marks values to be "registered" in pipeline registers.

Ah I see... yes, that's a nice abstraction. I agree that `register` fits better than `yield`.

> each stage should either read iter_args or values registered by any previous stage. I think that would provide some guardrails to make how things feed forward and back explicit, while still admitting multi-cycle operations. Thoughts?

Sounds reasonable to me, and would reduce clutter in the IR even without MCOs: When a value is used in multiple pipeline stages, then the lowering can figure out the best approach to implement the registered values, e.g. by instantiating shift register primitives to bridge longer use/def distances.

> One interesting case Morten brought up is what happens when iter_args feedback into the middle of a pipeline.

👍, I'd say that this IR should neither assume that the iter_args are always fed into stage 0 (otherwise you could just put a special terminator into stage "II-1", to yield the iter_args.), nor that all iter_args are registered from/fed into the same stage.

**Comment by mikeurbach** [comment]
I just pushed an update here, which covered a few things:

* Renamed `staticlogic.pipeline.yield` -> `staticlogic.pipeline.register`
* Documented that `condition` region must be "combinational" (in terms of certain `Standard` dialect ops)
* Documented that there is a relationship between `iter_args` def/use and II

I didn't make II an attribute yet, because I think it would be nice to use whatever we settle on in #1783.

I also pushed  new test, `test5`. This came from the following:

```C
void test5(int a[10]) {
  for (int i = 2; i < 10; ++i) {
    a[i] = a[i - 2] + a[i - 1];
  }
}
```

This was me trying to test II=2. First off, please sanity check, can we achieve II=2 for this loop? I'm assuming memory reads and writes take 1 cycle, and addition/subtraction is combinational.

Given the above, I compiled with Polygeist and turned the resulting SCF into the following pipeline:

```mlir
func @test5(%arg0: memref<?xi32>) {
  %c1 = constant 1 : index
  %c2 = constant 2 : index
  %c10 = constant 10 : index                                                                                                                                                                                   
  staticlogic.pipeline.while iter_args(%arg1 = %c2) : (index) -> () {
    %0 = cmpi ult, %arg1, %c10 : index
    staticlogic.pipeline.register %0 : i1
  } do {
    %0 = staticlogic.pipeline.stage  {
      %2 = subi %arg1, %c2 : index
      %3 = memref.load %arg0[%2] : memref<?xi32>
      staticlogic.pipeline.register %3 : i32
    } : i32
    %1:3 = staticlogic.pipeline.stage  {
      %2 = subi %arg1, %c1 : index
      %3 = memref.load %arg0[%2] : memref<?xi32>
      %4 = addi %arg1, %c1 : index
      staticlogic.pipeline.register %3, %arg1, %4 : i32, index, index
    } : i32, index, index
    staticlogic.pipeline.stage  {
      %2 = addi %0, %1#0 : i32
      memref.store %2, %arg0[%1#1] : memref<?xi32>
      staticlogic.pipeline.register
    }
    staticlogic.pipeline.terminator iter_args(%1#2), results() : (index) -> ()
  }
  return
}
```

Some things to note about the above:

* It isn't captured in an attribute, but II is intended to be 2
* Unlike the other examples, I've sunk the induction variable update and register to the second stage
* Because the second stage is computing the next value of the induction variable, and the third stage needs it, I've explicitly passed its value forward through a register in `%1#1`
* The third stage accesses registered results from both the first and second stage, demonstrating the ability to reach back further than the previous stage, which also came up when discussing multiple-cycle operations

I _think_ this is the way I'd want to represent such a loop, and it should satisfy the documented semantics. Specifically, this part:

> The maximum distance in stages between where an `iter_arg` is used and where that `iter_arg` is registered must be less than the loop's initiation interval. For example, with II=1, each `iter_arg` must be used and registered in the same stage.

The earliest use of `%arg1`, the induction variable, is in stage 1. It is registered in stage 2, so the distance is 1. This is less than the initiation interval of 2.

I played with several options in my head before settling on this. To convince myself, I tried to sketch out what memory accesses occur on each cycle, and what the pipeline would look like, according to my mental model:

IMG-0287

If my mental compiler is correct, this should ensure the `store` to a given address happens the cycle before it is next `load`ed from, which should match the C semantics.

What do you all think of this? Is this a reasonable way to represent such a loop?

**Comment by mortbopet** [comment]
The translation looks good to me. What i'm wondering about is at what level are we trying to be in charge of analyzing the loop here.
In my mind, analysis of cross-iteration memory dependencies is carried out at an affine level before we try pipeline scheduling our loop. This info is then factored in when scheduling the loop, annotated as a minimum II or dependency distance relations. 
At the pipeline level we can create our own constraints on II through how we place stages and observe the distance between use-def's of the iter_args, but memory dependencies should (in my mind) represent a fixed lower-bound for the II of the loop, generated by affine analysis.

**Comment by jopperm** [comment]
> I also pushed new test, `test5`. This came from the following:
> 
> ```c
> void test5(int a[10]) {
>   for (int i = 2; i < 10; ++i) {
>     a[i] = a[i - 2] + a[i - 1];
>   }
> }
> ```
> 
> This was me trying to test II=2. First off, please sanity check, can we achieve II=2 for this loop? I'm assuming memory reads and writes take 1 cycle, and addition/subtraction is combinational.

Yes, the recurrence-constrained minimum II is 2 here, induced by the recurrence: (dep distance in parantheses)
```
load a[i-1] -(0)-> store a[i] -(1)-> load a[i-1]
```

> * Because the second stage is computing the next value of the induction variable, and the third stage needs it, I've explicitly passed its value forward through a register in `%1#1`

Could we place the burden of figuring out the placement of registers on the lowering pass(es)? I think it'd be more intuitive if the SSA value representing the iter_arg could be used throughout the pipeline, i.e. before and after `stage(first use) + II-1`.

> What do you all think of this? Is this a reasonable way to represent such a loop?

Your mental compiler LGTM as well. I think that the pipeline op would be a valuable IR a) to capture the scheduling result, and b) to construct a hardware datapath and controller FSM from. However, I agree with @mortbopet that analysis and transformation opportunities are limited on this IR, due to the missing additional info/constraints that went into the scheduling problem. Hence, I'd see breaking down the conversion from SCF/Affine to FSMD into more manageable parts as the main benefit of this IR.

**Comment by mikeurbach** [comment]
Thanks for the feedback. Regarding affine analysis and memory dependence, I'm in complete agreement. I'm actually working on a pass right now to do the affine analysis and construct a scheduling problem at the level of affine loops. Then I'm going to work out how to take the result of that and create one of these pipeline ops. If there's anything useful from that analysis to carry along into the pipeline, I can definitely enhance the op with that information.

**Comment by mikeurbach** [comment]
> I think it'd be more intuitive if the SSA value representing the iter_arg could be used throughout the pipeline, i.e. before and after stage(first use) + II-1

This is something I wrestled with yesterday. I agree that making this implicit can make sense if it is well documented in the op's semantics. My struggle was that I also intended for registers to be made explicit at this level for feed forward values. Perhaps a clear line in the sand is to say feed forward values and definitions of iter_args must explicitly be operands to `staticlogic.pipeline.register`, but the uses of iter_args can have their registers be implicit, and handled by lowering. Given the information captured here, I think this is reasonable (especially if minimum II and/or dependence distances are made explicit).

At some point, I think we need to just land this op, and as we start working on lowering into/out of this representation, we should feel free to adjust the IR as we learn what is most beneficial. I'll make one last tweak to remove the `%1#1` register and leave it implicit, then submit this for a round of review.

File: include/circt/Dialect/StaticLogic/StaticLogic.td [hunk 2]
```diff
@@ -72,4 +72,159 @@ def ReturnOp : Op<StaticLogic_Dialect, "return", [Terminator]> {
...
+
+def PipelineTerminatorOp : Op<StaticLogic_Dialect, "pipeline.terminator",
+    [HasParent<"PipelineWhileOp">, Terminator, AttrSizedOperandSegments]> {
+  let summary = "StaticLogic dialect pipeline terminator.";
+  let description = [{
+    The `staticlogic.pipeline.terminator` operation represents the terminator of
+    a `staticlogic.pipeline.while`.
+
+    The `results` section accepts a variadic list of values which become the
+    pipeline’s return values. These must be results of the last stage, and their
```
Comment (jopperm) [inline include/circt/Dialect/StaticLogic/StaticLogic.td:206]: The last-stage-constraint has been lifted, hasn't it?

**Review by jopperm — APPROVED** [review]
> I think we need to just land this op, and as we start working on lowering into/out of this representation, we should feel free to adjust the IR as we learn what is most beneficial.

👍

File: include/circt/Dialect/StaticLogic/StaticLogic.td (hunk 2 above)
Comment (mikeurbach) [reply include/circt/Dialect/StaticLogic/StaticLogic.td:206]: Hmm for `staticlogic.pipeline.terminator`, I was intending for results to always come from the final stage (I did miss a verifier for that). I lifted the constraint that stages must always access results from the stage before them, but for the outputs of the pipeline itself, shouldn't these always come from the last/final stage? I haven't really thought through what it would look like to output intermediate stages' results.

Comment (jopperm) [reply include/circt/Dialect/StaticLogic/StaticLogic.td:206]: > I did miss a verifier for that.

Ah ok, that led me to believe this was just an oversight in the docs.

Anyways, my thinking was, once a value is marked as registered, the lowering can preserve it until the end of the pipeline automatically. Allowing results to be computed in earlier stages could be useful if the pipeline returns multiple values that don't necessarily become available in the same stage, or if some stateful ops (e.g. a `memref.store`) occupy the last stage.

Comment (mikeurbach) [reply include/circt/Dialect/StaticLogic/StaticLogic.td]: Updated the wording to note the "final" stage for pipeline output results, and added a verifier to enforce this. If there is a use-case to allow intermediate stages' results to be output, we can walk this back, but I think this is a good place to start.

Comment (mikeurbach) [reply include/circt/Dialect/StaticLogic/StaticLogic.td]: Ah I had missed your previous comment @jopperm.

> Allowing results to be computed in earlier stages could be useful if the pipeline returns multiple values that don't necessarily become available in the same stage, or if some stateful ops (e.g. a memref.store) occupy the last stage.

I think this does make sense, as long as we document that the lowering must take this into account. I'll walk back my last change and update the docs.

**Comment by mikeurbach** [comment]
I've put back the ability for any stage's results to be outputs. I've also added docs for two behaviors that are implicit, and need to be handled in lowering:

* If a stage accesses an `iter_arg` after the stage in which it is defined, it is up to lowering passes to preserve this value until the last stage that needs it.
* The results need not be defined in the final stage, and it is up to lowering passes to preserve these values until the final stage is complete.

I think this is a good starting point, but I'll leave the PR till early next week in case there is any further feedback.

File: lib/Dialect/StaticLogic/StaticLogicOps.cpp [hunk 3]
```diff
@@ -11,13 +11,194 @@
...
+
+    if (!isa<AddIOp, AndOp, BitcastOp, CmpIOp, ConstantOp, IndexCastOp, MulIOp,
+             OrOp, SelectOp, ShiftLeftOp, SignExtendIOp, SignedCeilDivIOp,
+             SignedDivIOp, SignedFloorDivIOp, SignedRemIOp, SignedShiftRightOp,
+             SubIOp, TruncateIOp, UnsignedDivIOp, UnsignedRemIOp,
+             UnsignedShiftRightOp, XOrOp, ZeroExtendIOp>(op)) {
+      nonCombinational = op;
+      return WalkResult::interrupt();
+    }
+
```
Comment (mortbopet) [inline lib/Dialect/StaticLogic/StaticLogicOps.cpp:100]: nit: no newline

File: include/circt/Dialect/StaticLogic/StaticLogic.td [hunk 4]
```diff
@@ -72,4 +72,163 @@ def ReturnOp : Op<StaticLogic_Dialect, "return", [Terminator]> {
...
+}
+
+def PipelineStageOp : Op<StaticLogic_Dialect, "pipeline.stage",
+    [HasParent<"PipelineWhileOp">]> {
+  let summary = "StaticLogic dialect pipeline stage.";
+  let description = [{
+    This operation has a single-block region which dictates the operations that
+    may occur concurrently.
+
+    It may have an optional `when` predicate, in which case the stage only
```
[the remark this answers is not in the document]
Comment (mortbopet) [reply include/circt/Dialect/StaticLogic/StaticLogic.td]: > ... I can add some wording that this when predicate is an additional gate on top of the condition.  

In that case, it wouldn't be lowering-defined, correct?

Hmm, I'm trying to imagine a situation where it is _not_ the case that the `condition` will be and-ed with the `when` predicate... Whether lowering defined or not, a comment on it should be sufficient for now :+1:.
